src/background.c

```diff
@@ -0,0 +1,817 @@
+                                              name,
+                                              NULL,
+                                              notify_background_done,
+                                              ddata);
+
+  g_assert (idata->handle == NULL);
+  idata->handle = handle;
+  idata->notified = TRUE;
+}
+
+static void
+check_background_apps (void)
+{
+  g_autoptr(GVariant) perms = NULL;
+  g_autoptr(GHashTable) app_states = NULL;
+  g_autoptr(GPtrArray) instances = NULL;
+  int i;
+  static int stamp;
+
+  app_states = get_app_states ();
+  if (app_states == NULL)
+    return;
+
+  g_debug ("Checking background permissions");
+
+  perms = get_all_permissions ();
+  instances = flatpak_instance_get_all ();
+
+  stamp++;
+
+  for (i = 0; i < instances->len; i++)
+    {
+      FlatpakInstance *instance = g_ptr_array_index (instances, i);
+      const char *id;
+      const char *app_id;
+      InstanceData *idata;
+      Permission permission;
+      const char *state_names[] = { "background", "running", "active" };
+      gboolean is_new = FALSE;
+
+      if (!flatpak_instance_is_running (instance))
+        continue;
+
+      id = flatpak_instance_get_id (instance);
+      app_id = flatpak_instance_get_app (instance);
+      idata = g_hash_table_lookup (applications, id);
+
+      if (!idata)
+        {
+          is_new = TRUE;
+          idata = g_new0 (InstanceData, 1);
+          idata->instance = g_object_ref (instance);
+          g_hash_table_insert (applications, g_strdup (id), idata);
+        }
+
+      idata->stamp = stamp;
+      idata->state = get_one_app_state (app_id, app_states);
+
+      g_debug ("App %s is %s", app_id, state_names[idata->state]);
+
+      permission = get_one_permission (app_id, perms);
+
+      if (idata->permission != permission)
+        {
+          /* Notify again if permissions change */
+          idata->permission = permission;
+          idata->notified = FALSE;
+        }
+
+      /* If the app is not in the list yet, add it,
+       * but don't notify yet - this gives apps some
+       * leeway to get their window app. If it is still
+       * in the background next time around, we'll proceed
+       * to the next step.
+       */
+      if (idata->state != BACKGROUND || idata->notified || is_new)
+        continue;
+
+      switch (idata->permission)
+        {
+        case NO:
+          kill_instance (id);
+          idata->stamp = 0;
+          break;
+
+        case ASK:
+        case UNSET:
+          send_notification (idata);
+          break;
+
+        case YES:
+        default:
+          break;
+        }
+    }
+
+  remove_outdated_instances (stamp);
+}
+
+static gpointer
+background_monitor (gpointer data)
+{
+  applications = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                        g_free, instance_data_free);
+  while (1)
+    {
+      check_background_apps ();
+      sleep (20);
```

Comment:
The docs say each minute, but this is 20 sec. I think every minute is a better value, as there is no need for this to be polling so often.